File: gradual_gen/gens/gen3/packages/physics/src/prng.ts
/**
 * PCG Random Number Generator (TypeScript implementation)
 * 
 * This implementation matches the Rust PCG-XSH-RR exactly for determinism.
 * Uses BigInt for 64-bit arithmetic.
 */

const PCG_MULTIPLIER = 6364136223846793005n;

/**
 * PCG-XSH-RR random number generator
 * 
 * Produces identical sequences to the Rust implementation.
 */
export class Pcg32 {
    private state: bigint;
    private inc: bigint;

    /**
     * Create a new PCG with seed and stream
     */
    constructor(seed: bigint = 0n, stream: bigint = 0n) {
        this.state = 0n;
        this.inc = (stream << 1n) | 1n; // Ensure odd

        // Advance state once
        this.nextU32();

        // Mix in the seed
        this.state = (this.state + seed) & 0xFFFFFFFFFFFFFFFFn;

        // Advance again
        this.nextU32();
    }

    /**
     * Create from saved state
     */
    static fromState(state: bigint, inc: bigint): Pcg32 {
        const rng = new Pcg32(0n, 0n);
        rng.state = state;
        rng.inc = inc;
        return rng;
    }

    /**
     * Get current state for serialization
     */
    getState(): [bigint, bigint] {
        return [this.state, this.inc];
    }

    /**
     * Generate next 32-bit unsigned integer
     */
    nextU32(): number {
        const oldState = this.state;

        // Advance internal state
        this.state = ((oldState * PCG_MULTIPLIER) + this.inc) & 0xFFFFFFFFFFFFFFFFn;

        // Calculate output function (XSH-RR)
        const xorshifted = Number(((oldState >> 18n) ^ oldState) >> 27n) >>> 0;
        const rot = Number(oldState >> 59n);

        // Rotate right
        return ((xorshifted >>> rot) | (xorshifted << ((-rot) & 31))) >>> 0;
    }

    /**
     * Generate a 64-bit unsigned integer
     */
    nextU64(): bigint {
        const high = BigInt(this.nextU32());
        const low = BigInt(this.nextU32());
        return (high << 32n) | low;
    }

    /**
     * Generate a float in [0, 1)
     */
    nextF64(): number {
        // Use 53 bits for full double precision mantissa
        const bits = this.nextU64() >> 11n;
        return Number(bits) / (1 << 53);
    }

    /**
     * Generate a float in [min, max)
     */
    nextF64Range(min: number, max: number): number {
        return min + (max - min) * this.nextF64();
    }

    /**
     * Generate an integer in [0, bound)
     */
    nextU32Bounded(bound: number): number {
        if (bound <= 0) return 0;

        // Lemire's method for unbiased bounded random
        let x = this.nextU32();
        let m = BigInt(x) * BigInt(bound);
        let l = Number(m & 0xFFFFFFFFn);

        if (l < bound) {
            const threshold = ((-bound) >>> 0) % bound;
            while (l < threshold) {
                x = this.nextU32();
                m = BigInt(x) * BigInt(bound);
                l = Number(m & 0xFFFFFFFFn);
            }
        }

        return Number(m >> 32n);
    }

    /**
     * Generate a random boolean
     */
    nextBool(): boolean {
        return (this.nextU32() & 1) === 1;
    }

    /**
     * Generate a normally distributed random number (Box-Muller transform)
     */
    nextGaussian(mean: number = 0, stdDev: number = 1): number {
        let u1 = this.nextF64();
        const u2 = this.nextF64();

        // Avoid log(0)
        if (u1 < Number.EPSILON) u1 = Number.EPSILON;

        const z0 = Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
        return mean + z0 * stdDev;
    }

    /**
     * Generate a random unit vector (uniform on sphere surface)
     */
    nextUnitVector(): [number, number, number] {
        while (true) {
            const x = this.nextF64Range(-1, 1);
            const y = this.nextF64Range(-1, 1);
            const z = this.nextF64Range(-1, 1);

            const lenSq = x * x + y * y + z * z;

            if (lenSq > Number.EPSILON && lenSq <= 1) {
                const len = Math.sqrt(lenSq);
                return [x / len, y / len, z / len];
            }
        }
    }

    /**
     * Shuffle an array in place (Fisher-Yates)
     */
    shuffle<T>(array: T[]): void {
        for (let i = array.length - 1; i > 0; i--) {
            const j = this.nextU32Bounded(i + 1);
            [array[i], array[j]] = [array[j], array[i]];
        }
    }

    /**
     * Choose a random element from an array
     */
    choose<T>(array: T[]): T | undefined {
        if (array.length === 0) return undefined;
        return array[this.nextU32Bounded(array.length)];
    }
}

// Test determinism
if (typeof process !== 'undefined' && process.env.NODE_ENV === 'test') {
    const rng1 = new Pcg32(42n, 54n);
    const rng2 = new Pcg32(42n, 54n);

    for (let i = 0; i < 1000; i++) {
        const v1 = rng1.nextU32();
        const v2 = rng2.nextU32();
        if (v1 !== v2) {
            throw new Error(`PRNG determinism failed at iteration ${i}`);
        }
    }
}

File: gradual_gen/gens/gen3/packages/physics/src/prng.rs
//! PCG (Permuted Congruential Generator) random number generator
//!
//! Deterministic PRNG that produces identical sequences across Rust and TypeScript.
//! Implementation follows PCG-XSH-RR variant (32-bit output, 64-bit state).

use serde::{Deserialize, Serialize};

/// PCG-XSH-RR random number generator
///
/// This implementation is designed to be deterministic and reproducible
/// across different platforms (Rust native, WASM, JavaScript port).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Pcg32 {
    /// Internal state
    state: u64,
    /// Increment (stream selector), must be odd
    inc: u64,
}

impl Pcg32 {
    /// Default multiplier for PCG
    const MULTIPLIER: u64 = 6364136223846793005;

    /// Create a new PCG with the given seed and stream
    ///
    /// # Arguments
    /// * `seed` - Initial seed value
    /// * `stream` - Stream selector (will be made odd internally)
    pub fn new(seed: u64, stream: u64) -> Self {
        let mut rng = Self {
            state: 0,
            inc: (stream << 1) | 1, // Ensure odd
        };
        // Advance state once
        rng.next_u32();
        // Mix in the seed
        rng.state = rng.state.wrapping_add(seed);
        // Advance again
        rng.next_u32();
        rng
    }

    /// Create a PCG seeded from the current time (non-deterministic)
    #[cfg(not(target_arch = "wasm32"))]
    pub fn from_entropy() -> Self {
        use std::time::{SystemTime, UNIX_EPOCH};
        let seed = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_nanos() as u64)
            .unwrap_or(0x853c49e6748fea9b);
        Self::new(seed, 0xda3e39cb94b95bdb)
    }

    /// Create a PCG seeded from entropy (WASM version)
    #[cfg(target_arch = "wasm32")]
    pub fn from_entropy() -> Self {
        let mut seed_bytes = [0u8; 8];
        getrandom::getrandom(&mut seed_bytes).unwrap_or_default();
        let seed = u64::from_le_bytes(seed_bytes);
        Self::new(seed, 0xda3e39cb94b95bdb)
    }

    /// Create a PCG from serialized state (for snapshot restore)
    pub fn from_state(state: u64, inc: u64) -> Self {
        Self { state, inc }
    }

    /// Get the current state for serialization
    pub fn get_state(&self) -> (u64, u64) {
        (self.state, self.inc)
    }

    /// Generate the next 32-bit unsigned integer
    pub fn next_u32(&mut self) -> u32 {
        let old_state = self.state;

        // Advance internal state
        self.state = old_state
            .wrapping_mul(Self::MULTIPLIER)
            .wrapping_add(self.inc);

        // Calculate output function (XSH-RR)
        let xorshifted = (((old_state >> 18) ^ old_state) >> 27) as u32;
        let rot = (old_state >> 59) as u32;
        xorshifted.rotate_right(rot)
    }

    /// Generate a 64-bit unsigned integer (two 32-bit values combined)
    pub fn next_u64(&mut self) -> u64 {
        let high = self.next_u32() as u64;
        let low = self.next_u32() as u64;
        (high << 32) | low
    }

    /// Generate a floating-point number in [0, 1)
    pub fn next_f64(&mut self) -> f64 {
        // Use 53 bits for full double precision mantissa
        let bits = self.next_u64() >> 11;
        bits as f64 * (1.0 / (1u64 << 53) as f64)
    }

    /// Generate a floating-point number in [min, max)
    pub fn next_f64_range(&mut self, min: f64, max: f64) -> f64 {
        min + (max - min) * self.next_f64()
    }

    /// Generate an integer in [0, bound)
    pub fn next_u32_bounded(&mut self, bound: u32) -> u32 {
        // Lemire's method for unbiased bounded random
        let mut x = self.next_u32();
        let mut m = (x as u64).wrapping_mul(bound as u64);
        let mut l = m as u32;

        if l < bound {
            let threshold = bound.wrapping_neg() % bound;
            while l < threshold {
                x = self.next_u32();
                m = (x as u64).wrapping_mul(bound as u64);
                l = m as u32;
            }
        }

        (m >> 32) as u32
    }

    /// Generate a random boolean
    pub fn next_bool(&mut self) -> bool {
        (self.next_u32() & 1) == 1
    }

    /// Generate a normally distributed random number (Box-Muller transform)
    pub fn next_gaussian(&mut self, mean: f64, std_dev: f64) -> f64 {
        // Box-Muller transform
        let u1 = self.next_f64();
        let u2 = self.next_f64();

        // Avoid log(0)
        let u1 = if u1 < f64::EPSILON { f64::EPSILON } else { u1 };

        let z0 = (-2.0 * u1.ln()).sqrt() * (2.0 * std::f64::consts::PI * u2).cos();
        mean + z0 * std_dev
    }

    /// Generate a random unit vector (uniform on sphere surface)
    pub fn next_unit_vector(&mut self) -> [f64; 3] {
        // Use rejection sampling for uniform distribution on sphere
        loop {
            let x = self.next_f64_range(-1.0, 1.0);
            let y = self.next_f64_range(-1.0, 1.0);
            let z = self.next_f64_range(-1.0, 1.0);

            let len_sq = x * x + y * y + z * z;

            if len_sq > f64::EPSILON && len_sq <= 1.0 {
                let len = len_sq.sqrt();
                return [x / len, y / len, z / len];
            }
        }
    }

    /// Shuffle a slice in place (Fisher-Yates)
    pub fn shuffle<T>(&mut self, slice: &mut [T]) {
        let len = slice.len();
        if len <= 1 {
            return;
        }

        for i in (1..len).rev() {
            let j = self.next_u32_bounded((i + 1) as u32) as usize;
            slice.swap(i, j);
        }
    }
}

impl Default for Pcg32 {
    fn default() -> Self {
        // Default seed for reproducible behavior
        Self::new(0x853c49e6748fea9b, 0xda3e39cb94b95bdb)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_determinism() {
        let mut rng1 = Pcg32::new(12345, 67890);
        let mut rng2 = Pcg32::new(12345, 67890);

        for _ in 0..1000 {
            assert_eq!(rng1.next_u32(), rng2.next_u32());
        }
    }

    #[test]
    fn test_known_sequence() {
        // Known sequence for seed=42, stream=54
        let mut rng = Pcg32::new(42, 54);

        // These values should be consistent across implementations
        let vals: Vec<u32> = (0..5).map(|_| rng.next_u32()).collect();

        // First value should always be deterministic for this seed
        assert_eq!(vals[0], 2707161783);
    }

    #[test]
    fn test_f64_range() {
        let mut rng = Pcg32::new(12345, 0);

        for _ in 0..10000 {
            let val = rng.next_f64();
            assert!(val >= 0.0 && val < 1.0);
        }
    }

    #[test]
    fn test_bounded() {
        let mut rng = Pcg32::new(12345, 0);
        let bound = 100u32;

        for _ in 0..10000 {
            let val = rng.next_u32_bounded(bound);
            assert!(val < bound);
        }
    }

    #[test]
    fn test_state_restore() {
        let mut rng = Pcg32::new(12345, 67890);

        // Generate some values
        for _ in 0..100 {
            rng.next_u32();
        }

        // Save state
        let (state, inc) = rng.get_state();

        // Generate more values
        let expected: Vec<u32> = (0..10).map(|_| rng.next_u32()).collect();

        // Restore state
        let mut rng2 = Pcg32::from_state(state, inc);

        // Should produce same values
        let actual: Vec<u32> = (0..10).map(|_| rng2.next_u32()).collect();

        assert_eq!(expected, actual);
    }

    #[test]
    fn test_unit_vector() {
        let mut rng = Pcg32::new(12345, 0);

        for _ in 0..1000 {
            let v = rng.next_unit_vector();
            let len_sq = v[0] * v[0] + v[1] * v[1] + v[2] * v[2];
            assert!((len_sq - 1.0).abs() < 1e-10);
        }
    }
}
